clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2164,7 +2160,11 @@ static void printArgument(ASTContext &Ctx, raw_ostream &ArgOS,
     TypePolicy.SuppressTypedefs = true;
     TypePolicy.SuppressTagKeyword = true;
     QualType T = Arg.getAsType();
-    ArgOS << getKernelNameTypeString(T, Ctx, TypePolicy);
+    if (T->isEnumeralType())
```

Comment:
Hmm... i would expect getKernelNameTypeString to just return the right value for the enum instead of needing ot handle it here.

Reply:
This happens where there is a call to getKernelNameTypeString from within the printTemplateArguments. Without this the kernelname-enum.cpp in dummy_functor_7<no_namespace_int> was not generating the "::". I initially put this in getKernelNameTypeString when RD is null, but that broke some other test cases where the "::" were added when they shouldn't. Let me look more into it.

Reply:
Still unresolved.  That said, I DO think that getKernelNameTypeString should just 'do the right thing' here, rather than having this special case.

Reply:
This will have to be special cased if not here then in getKernelNameTypeString. It would be the case where RD is not a record and T is an Enum. Will post the code to see if you agree with it.